Point at `const` definition when used instead of a binding in a `let` statement

estebank opened the pull request:
Modify `PatKind::InlineConstant` to be `ExpandedConstant` standing in not only for inline `const` blocks but also for `const` items. This allows us to track named `const`s used in patterns when the pattern is a single binding. When we detect that there is a refutable pattern involving a `const` that could have been a binding instead, we point at the `const` item, and suggest renaming. We do this for both `let` bindings and `match` expressions missing a catch-all arm if there's at least one single binding pattern referenced.

After:

```
error[E0005]: refutable pattern in local binding
  --> $DIR/bad-pattern.rs:19:13
   |
LL |     const PAT: u32 = 0;
   |     -------------- missing patterns are not covered because `PAT` is interpreted as a constant pattern, not a new variable
...
LL |         let PAT = v1;
   |             ^^^ pattern `1_u32..=u32::MAX` not covered
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
   = note: the matched value is of type `u32`
help: introduce a variable instead
   |
LL |         let PAT_var = v1;
   |             ~~~~~~~
```

Before:

```
error[E0005]: refutable pattern in local binding
  --> $DIR/bad-pattern.rs:19:13
   |
LL |         let PAT = v1;
   |             ^^^
   |             |
   |             pattern `1_u32..=u32::MAX` not covered
   |             missing patterns are not covered because `PAT` is interpreted as a constant pattern, not a new variable
   |             help: introduce a variable instead: `PAT_var`
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
   = note: the matched value is of type `u32`
```

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r\? <reviewer name> (with the `\` removed)
-->
<!-- homu-ignore:end -->

CC #132582.

estebank committed:
Point at `const` definition when used instead of a binding in a `let` statement

After:

```
error[E0005]: refutable pattern in local binding
  --> $DIR/bad-pattern.rs:19:13
   |
LL |     const PAT: u32 = 0;
   |     -------------- missing patterns are not covered because `PAT` is interpreted as a constant pattern, not a new variable
...
LL |         let PAT = v1;
   |             ^^^
   |             |
   |             pattern `1_u32..=u32::MAX` not covered
   |             help: introduce a variable instead: `PAT_var`
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
   = note: the matched value is of type `u32`
```

Before:

```
error[E0005]: refutable pattern in local binding
  --> $DIR/bad-pattern.rs:19:13
   |
LL |         let PAT = v1;
   |             ^^^
   |             |
   |             pattern `1_u32..=u32::MAX` not covered
   |             missing patterns are not covered because `PAT` is interpreted as a constant pattern, not a new variable
   |             help: introduce a variable instead: `PAT_var`
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
   = note: the matched value is of type `u32`
```

compiler-errors commented on compiler/rustc_mir_build/src/thir/pattern/mod.rs:
```diff
@@ -560,9 +562,15 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
             _ => return pat_from_kind(self.lower_variant_or_leaf(res, id, span, ty, vec![])),
         };
 
+        // HERE
```
?

compiler-errors reviewed (changes requested):
This PR needs a lot more explanation that details why we need to add new THIR, or better yet, it could use some more investigation towards how we can avoid needing to add THIR at all -- since this is for the diagnostic path. Like, why can't we just extract the def span of the const from the `UnevaluatedConst` that is inside the `mir::Const`?

Adding new THIR nodes seems like it's ripe for adding confusion to an already confusing area of the compiler, and this PR has no explanation or documentation other than a "before" and "after". I'm not certain if it's worth the overhead to add all this to the compiler to improve this diagnostic if this is the only way to do it.

compiler-errors commented:
Even if we need to end up plumbing this span separately, why not just add an `opt_def_span: Option<Span>` to the existing `Constant` patkind?

estebank commented:
> why can't we just extract the def span of the const from the `UnevaluatedConst` that is inside the `mir::Const`?

Because by the time the error occurs, it is no longer an `UnevaluatedConst`.

> and this PR has no explanation or documentation other than a "before" and "after".

It's providing additional context to the output from what we have today. Today the output can be quite confusing, particularly if the constant is defined far away [with a non-standard identifier](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=98180dc81bd5925ecd473ca866c5581f), and [it can be even worse if it happens due to a re-export that renames it](https://github.com/rust-lang/rust/pull/132708/files#diff-abd52265f9ba2cfd08e173caa3e6a27f06c6bbe038e489e6d373657cff3f1cf0R17-R31). The error as it is today takes me a triple take to understand what it was telling me about.

> I'm not certain if it's worth the overhead to add all this to the compiler to improve this diagnostic if this is the only way to do it.

I'm more than happy to hear about alternative ways to do this. I disagree that the output as is today is reasonable. I'd go as far as saying it is inescrutable for anyone who doesn't yet know about the handling of `const`s in patterns or that `let` bindings are patterns to begin with.

> Even if we need to end up plumbing this span separately, why not just add an opt_def_span: Option<Span> to the existing Constant patkind?

I'm ok with that alternative.

estebank committed:
Fold `PatKind::NamedConstant` into `PatKind::Constant`

compiler-errors commented:
All of that information would've been awesome to include in the PR description.

I hope you don't take this in the wrong way, but it's a bit annoying from a reviewer's perspective (and kinda leads to additional work on the reviewer) to have to try to deduce the reasoning that went into this approach from basically no information, or to have to leave a bunch of confused questions and go back and forth on a semi-synchronous github convo. 

This is especially true when it touches something very important like THIR, as opposed to diagnostics PRs that generally stay out of the way of the good path. Divergence in the way we handle `PatKind`s can lead to bugs or ICEs.

> I disagree that the output as is today is reasonable.

I think you're misinterpreting me. What I mean is that I'm not certain if the approach is worth the overhead. I'm not saying that the output today is *reasonable* -- what I'm saying is that I'm not certain if we should land this approach, as opposed to deferring fixing this in favor of investigating a better solution for this.

> >  Even if we need to end up plumbing this span separately, why not just add an opt_def_span: Option to the existing Constant patkind?
>
> I'm ok with that alternative.

Well, if you plumb in the `Option<DefId>` of the unevaluated const, then you can avoid the `span_to_snippet` call in the suggestion producing code, because we can get the span and item name of the const at that point from the `TyCtxt`. It's probably best to do that if we're plumbing more information into the `Const` pattern-kind, since then we make things cleaner than we found them :)

compiler-errors commented on compiler/rustc_mir_build/src/thir/pattern/check_match.rs:
```diff
@@ -668,8 +668,21 @@ impl<'p, 'tcx> MatchVisitor<'p, 'tcx> {
+            subpattern: box Pat { kind: PatKind::Constant { opt_def: Some(def_id), .. }, .. },
+            ..
+        } = pat.kind
+            && let Ok(snippet) = self.tcx.sess.source_map().span_to_snippet(pat.span)
```
Ah, you read my mind with the `DefId` -- but we can take it further and use the `tcx.item_name` here to get the name of the constant. This will regress reexported constants, but is more resilient IMO and is worth the overhead.

compiler-errors replied:
Specifically, this pat span will not possibly be tied to arbitrary source code, in `pat_macro!(/* uwu */)`.

estebank committed:
Use `item_name` instead of a span snippet when talking about const pattern

Nadrieril commented:
This is a pleasant diagnostic improvement. Am I correct to understand that this only catches constants of primitive type? Is there a way to catch also 
```rust
const PAT: (u32, u32) = ...;
let PAT = ...;
```
?

estebank commented:
@Nadrieril for that case, the output is

```
error[E0005]: refutable pattern in local binding
 --> f702.rs:9:9
  |
9 |     let PAT = (1, 1); 
  |         ^^^ pattern `(1_u32..=u32::MAX, _)` not covered
  |
  = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
  = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
  = note: the matched value is of type `(u32, u32)`
help: you might want to use `if let` to ignore the variant that isn't matched
  |
9 |     if let PAT = (1, 1) { todo!() }; 
  |     ++                  +++++++++++
```

Looking at the error it seems it is because it is checking the sub-pattern for the first tuple item. I can take a look to see if we can make it work for tuples (and arrays, and structs). For enums we already have reasonable output before this PR:

```
error[E0005]: refutable pattern in local binding
  --> f702.rs:15:9
   |
15 |     let PAT2 = E::B;
   |         ^^^^ pattern `E::B` not covered
   |
   = note: `let` bindings require an "irrefutable pattern", like a `struct` or an `enum` with only one variant
   = note: for more information, visit https://doc.rust-lang.org/book/ch18-02-refutability.html
note: `E` defined here
  --> f702.rs:5:6
   |
5  | enum E {
   |      ^
6  |     A,
7  |     B,
   |     - not covered
   = note: the matched value is of type `E`
help: you might want to use `if let` to ignore the variant that isn't matched
   |
15 |     if let PAT2 = E::B { todo!() };
   |     ++                 +++++++++++
```

Nadrieril commented:
It would be nice to have it point at the definition of the constant in these cases too. Once we get to `PatKind` it's too late however so we'd have to get the info before. Hm, unless we want to add a `PatKind::NamedConst { subpat: Pat, id: DefId }` variant I guess, which isn't unreasonable given that we already have `PatKind::InlineConst` that looks very similar, no? Or did we remove that one

estebank commented:
@Nadrieril  We do have `PatKind::InlineConst`. Adding `PatKind::NamedConst` would indeed work. I leave it to you and @compiler-errors to decide whether the additional maintenance cost is reasonable or not. I believe it is, given that I'm finding another place to use these for patterns where the intended fall-through binding is mixed up with a const name, so there are already at least two cases where we want this metadata, and not supporting anything beyond literals (and enums, that we already accounted for) would be a shame.

Nadrieril commented:
If we rename `InlineConstant` to `ExpandedConstant` and add a `is_inline: bool` field if needed, that should be no extra maintenance cost. I'm in favor of that, it's more principled that the current approach in this PR

estebank commented:
@Nadrieril there's one wrinkle there: `InlineConstant` stores a `LocalDefId`, but `const`s can come from other crates so it'd have to be a `DefId`.

estebank committed:
Point at const when intended binding fall-through pattern is a const

```
error[E0004]: non-exhaustive patterns: `i32::MIN..=3_i32` and `5_i32..=i32::MAX` not covered
  --> $DIR/intended-binding-pattern-is-const.rs:2:11
   |
LL |     match 1 {
   |           ^ patterns `i32::MIN..=3_i32` and `5_i32..=i32::MAX` not covered
LL |         x => {}
   |         - this pattern doesn't introduce a new catch-all binding, but rather pattern matches against the value of constant `x`
   |
   = note: the matched value is of type `i32`
note: constant `x` defined here
  --> $DIR/intended-binding-pattern-is-const.rs:7:5
   |
LL |     const x: i32 = 4;
   |     ^^^^^^^^^^^^
help: if you meant to introduce a binding, use a different name
   |
LL |         x_var => {}
   |          ++++
help: ensure that all possible cases are being handled by adding a match arm with a wildcard pattern, a match arm with multiple or-patterns as shown, or multiple match arms
   |
LL |         x => {}, i32::MIN..=3_i32 | 5_i32..=i32::MAX => todo!()
   |                ++++++++++++++++++++++++++++++++++++++++++++++++
```

Nadrieril commented:
We can store a DefId and add a comment "if `is_inline` is true, this def_id is local"

estebank committed:
Unify expanded constants and named constants in `PatKind`

estebank committed:
Make suggestion verbose

Nadrieril commented on compiler/rustc_middle/src/thir.rs:
```diff
@@ -788,12 +788,16 @@ pub enum PatKind<'tcx> {
         /// reference that can be used by unsafety checking to visit nested
-        /// unevaluated constants.
-        def: LocalDefId,
+        /// unevaluated constants. If the `DefId` doesn't correspond to a local
```
```suggestion
        /// unevaluated constants and for diagnostics. If the `DefId` doesn't correspond to a local
```

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/check_match.rs:
```diff
@@ -668,8 +668,22 @@ impl<'p, 'tcx> MatchVisitor<'p, 'tcx> {
+            // When we encounter a constant as the binding name, point at the `const` definition.
+            interpreted_as_const = Some(span);
+            interpreted_as_const_sugg = Some(InterpretedAsConst { span: pat.span, variable });
+        } else if let PatKind::Constant { .. }
         | PatKind::AscribeUserType {
             subpattern: box Pat { kind: PatKind::Constant { .. }, .. },
             ..
```
I think this `else` is redundant now: any named constant will give a `ExpandedConstant` so there's no need for this special-case that only applies to integers/strings and uses `span_to_snippet`.

estebank replied:
This branch is for [`let` bindings without an identifier (generally a literal)](https://github.com/rust-lang/rust/blob/81eef2d362a6f03db6f8928f82d94298d31eb81b/tests/ui/closures/2229_closure_analysis/bad-pattern.stderr#L14-L17) and we can't just remove it outright without losing output. We *can* remove the later inner `else` though (which might have been what you meant).

Nadrieril commented on compiler/rustc_mir_build/src/thir/pattern/mod.rs:
```diff
@@ -562,7 +569,20 @@ impl<'a, 'tcx> PatCtxt<'a, 'tcx> {
+        let pattern = if let hir::QPath::Resolved(None, path) = qpath
+            && path.segments.len() == 1
+        {
+            // We only want to mark constants when referenced as bare names that could have been
+            // new bindings if the `const` didn't exist.
```
Hm, I assumed we'd be indiscriminately recording all constant expansions and would filter later. I'm not sure which is best, but at least please mention this choice in the `PatKind` variant.

estebank replied:
Instead of only setting them in this case we could add another flag to the variant to track this and *not* suggest if the flag is set.

Nadrieril replied:
That would be an improvement imo. I would still prefer to avoid carrying a boolean in THIR for a single diagnostic suggestion tbh, depends how hard it is to do without the boolean.

Nadrieril replied:
My preference would be: `ExpandedConst { def_id, subpattern }` and do the `span_to_snippet(subpattern.span).chars().all(is_alphanumeric)` dance for diagnostics.

estebank replied:
Let me know what you think with the last commit.

Nadrieril replied:
That's what I had in mind, ty. Could you add the `ExpandedConstant` layer inside `const_to_pat` by any chance? To be sure we don't miss one

Nadrieril replied:
I looked at `const_to_pat`, it wouldn't be super clean to add it there. I'm therefore ok with the current way of doing it.

Nadrieril commented on tests/ui/consts/const_in_pattern/incomplete-slice.stderr:
```diff
@@ -3,8 +3,20 @@ error[E0004]: non-exhaustive patterns: `&[]` and `&[_, _, ..]` not covered
    |
 LL |     match &[][..] {
    |           ^^^^^^^ patterns `&[]` and `&[_, _, ..]` not covered
+LL |
+LL |         E_SL => {}
+   |         ---- this pattern doesn't introduce a new catch-all binding, but rather pattern matches against the value of constant `E_SL`
    |
    = note: the matched value is of type `&[E]`
+note: constant `E_SL` defined here
+  --> $DIR/incomplete-slice.rs:6:1
+   |
+LL | const E_SL: &[E] = &[E::A];
+   | ^^^^^^^^^^^^^^^^
+help: if you meant to introduce a binding, use a different name
+   |
+LL |         E_SL_var => {}
+   |             ++++
```
This is great!

Nadrieril commented:
In terms of additional complexity, this now seems perfectly fine. The one thing I have mixed feelings about is that this only adds an `ExpandedConstant` if the constant could be mistaken for an identifier. This is mixing diagnostics concerns with representation concerns; I'd much prefer if we always added `ExpandedConstant` to the output of `const_to_pat`.

estebank committed:
review comments

estebank committed:
review comments

- Remove check for "how many path segments is the pattern"
- Check before suggesting if the path has multiple path segments

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-18`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11826486123/job/32952623572) [(plain)](https://github.com/rust-lang/rust/commit/3f84747d651e46881997518f09e65543d919beb6/checks/32952623572/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
------
 > importing cache manifest from ghcr.io/rust-lang/rust-ci-cache:c32c805632780b5c1de330e3f44561b336c2efe163bc0990acb392390157a8e1d9f855d75914a239aa40c49d77f4a837247d05d2f8d46f554b98e1f46712a3e3:
------
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-18]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-18', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-18/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
---- [mir-opt] tests/mir-opt/simplify_dead_blocks.rs stdout ----

error: compilation failed!
status: exit status: 1
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/mir-opt/simplify_dead_blocks.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "-O" "-Copt-level=1" "-Zdump-mir=SimplifyCfg-after-unreachable-enum-branching" "-Zvalidate-mir" "-Zlint-mir" "-Zdump-mir-exclude-pass-number" "-Zmir-include-spans=false" "--crate-type=rlib" "-Zmir-opt-level=0" "-Zmir-enable-passes=+SimplifyCfg-after-unreachable-enum-branching" "-Zdump-mir-dir=/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/simplify_dead_blocks" "--emit" "mir" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt/simplify_dead_blocks" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers"
--- stderr -------------------------------
warning: unused variable: `deadRetblock1`
##[warning]  --> /checkout/tests/mir-opt/simplify_dead_blocks.rs:36:9
   |
   |
36 |         deadRetblock1 = {
   |         ^^^^^^^^^^^^^
   |
help: `deadRetblock1` is captured in macro and introduced a unused variable
   |
27 | /     mir! {
28 | |         {
29 | |             match x {
---
   |             ^^^
   |
   = help: maybe it is overwritten before being read?

error: Could not parse constant pattern, found: "ExpandedConstant { def_id: DefId(2:33362 ~ core[b4a2]::num::{impl#6}::MAX), is_inline: false, subpattern: Pat { ty: u8, span: /checkout/tests/mir-opt/simplify_dead_blocks.rs:32:17: 32:24 (#0), kind: Constant { value: Ty(u8, u8::MAX) } } }"
   |
32 |                 u8::MAX => unreachable,
   |                 ^^^^^^^

```

</details>

Nadrieril commented on compiler/rustc_middle/src/thir.rs:
```diff
@@ -788,12 +788,19 @@ pub enum PatKind<'tcx> {
-    /// Inline constant found while lowering a pattern.
-    InlineConstant {
-        /// [LocalDefId] of the constant, we need this so that we have a
+    /// Inline or named constant found while lowering a pattern.
+    ///
+    /// We only mark patterns referencing constants when they are bare names that could have been
+    /// new bindings if the `const` didn't exist.
```
```suggestion
    /// Pattern obtained by converting a constant (inline or named) to its pattern representation.
```

Nadrieril commented:
I'm ok with this now! It's quite a nice diagnostic improvement. I think you have to fix THIR parsing, and I left one suggestion about a comment.

estebank committed:
review comment: modify doc comment

estebank committed:
Account for `ExpandedConstant` in `parse_match`

Nadrieril commented:
Ty!

@bors r+

bors commented:
:pushpin: Commit 29acf8b422ab446fa4cd51fbb0e5a145f30c1cfc has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 29acf8b422ab446fa4cd51fbb0e5a145f30c1cfc -->
<!-- homu: {"type":"Approved","sha":"29acf8b422ab446fa4cd51fbb0e5a145f30c1cfc","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
